lahja/endpoint.py

```diff
@@ -129,6 +198,68 @@ def connect_no_wait(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> N
 
             self._process_item(item, config)
 
+    def _create_external_api(self, ipc_path: pathlib.Path) -> None:
+
+        receiver = EndpointConnector(self)
+
+        class ConnectorManager(BaseManager):
+            pass
+
+        ConnectorManager.register('get_connector', callable=lambda: receiver)  # type: ignore
+
+        manager = ConnectorManager(address=str(ipc_path))  # type: ignore
+        server = manager.get_server()   # type: ignore
+        threading.Thread(target=server.serve_forever, daemon=True).start()
+
+    def connect_to_endpoints_blocking(self, *endpoints: ConnectionConfig, timeout: int=30) -> None:
+        """
+        Connect to the given endpoints and block until the connection to every endpoint is
+        established. Raises a ``TimeoutError`` if connections do not become available within
+        ``timeout`` seconds (default 30 seconds).
+        """
+        self._throw_if_already_connected(*endpoints)
+        for endpoint in endpoints:
+            wait_for_path_blocking(endpoint.path, timeout)
+            self._connect_to_endpoint(endpoint)
+
+    async def connect_to_endpoints(self, *endpoints: ConnectionConfig) -> None:
+        """
+        Connect to the given endpoints and await until all connections are established.
+        """
+        self._throw_if_already_connected(*endpoints)
```

Comment:
nitpick: This works in `connect_to_endpoints_blocking` but there's a race condition in `connect_to_endpoints`. I think that something like this would succeed even though it shouldn't:

```python
asyncio.gather(
    endpoint.connect_to_endpoints([ConnectionConfig.from_name('hi', path="first_path.ipc")]),
    endpoint.connect_to_endpoints([ConnectionConfig.from_name('hi', path="second_path.ipc")])
)
```

Because `connect_to_endpoints` does the check but defers actually adding the endpoint there's a race condition where a second endpoint with the same name could get past the check before the first one is added.

Reply:
Yep, I also noticed this. In fact, if you look at the test, I left out the one for the `nowait` API because this one is currently affected by this. I *think* the problem here is that in the affected cases you might end up with the exception thrown in a background task which is kinda bad. But maybe duplicating the check there is still the right thing to do to at least not sweep it under the rug.